### PR TITLE
JSON format for Chakra ET

### PR DESCRIPTION
**Pull request by rvinaybharadwaj**
## Summary
- Added JSON format for Chakra ET
- To maintain backward compatibility, a wrapper is included that presents a common API for Protobuf and JSON

## Test Plan
- Added wrapperTests.cpp and test data

### CODE REVIEW

**Comment by rvinaybharadwaj** [comment]
recheck

**Comment by rvinaybharadwaj** [comment]
recheck

**Comment by rvinaybharadwaj** [comment]
recheck

**Comment by JoongunPark** [comment]
Hi @rvinaybharadwaj,

Thank you for the PR! I have a few requests:

Could you ensure that the code passes all checks (file formatting, tests, etc.)?
Could you provide step-by-step commands to help us reproduce the results?
What is the expected outcome, and how can we verify it?

Thanks again!

**Comment by rvinaybharadwaj** [comment]
> Hi @rvinaybharadwaj,
> 
> Thank you for the PR! I have a few requests:
> 
> 1. Could you ensure that the code passes all checks (file formatting, tests, etc.)?
> 2. Could you provide step-by-step commands to help us reproduce the results?
> 3. What is the expected outcome, and how can we verify it?  (Could you share us example input and output?)
> 
> Thanks again!

Hi @JoongunPark,

1. I have run clang-format on the code and the tests are breaking on not finding a module during the setup phase. So, I don't think it's an issue related to my code. And I am not sure why the CLA check is failing even though I am a member of MLCommons.

2. I have added [wrapper tests](https://github.com/rvinaybharadwaj/chakra/blob/jsonify/tests/feeder/wrapperTests.cpp), which are functionally equivalent to the [feeder tests](https://github.com/rvinaybharadwaj/chakra/blob/jsonify/tests/feeder/tests.cpp) and have included [test data](https://github.com/rvinaybharadwaj/chakra/blob/jsonify/tests/data/jsonData.tar.gz). You can run it through the existing google test framework.

3. The data in chakra.0.json is the same as the data in chakra.0.et You should be able to achieve the same result as the protobuf ET and the wrapper tests should all pass.

Let me know if you have any more questions.

**Comment by JoongunPark** [comment]
Hi @rvinaybharadwaj,

Thank you for the update!

1. It looks like this branch has new conflicts after the recent merge. Could you please resolve them?
2. For the cpp_linter, you might want to update the GitHub workflow to fix it. Here’s the link: [cpp_lint.yml](https://github.com/mlcommons/chakra/blob/73edb74facc20f08f69a478c25e6fc2d207d9846/.github/workflows/cpp_lint.yml#L4).
3. Regarding the CLA issue, I believe someone from MLCommons needs to add you. @tushar-krishna or @srinivas212 should be able to help with this.

I'll run some tests and share the results.

**Comment by rvinaybharadwaj** [comment]
> Hi @rvinaybharadwaj,
> 
> Thank you for the update!
> 
> 1. It looks like this branch has new conflicts after the recent merge. Could you please resolve them?
> 2. For the cpp_linter, you might want to update the GitHub workflow to fix it. Here’s the link: [cpp_lint.yml](https://github.com/mlcommons/chakra/blob/73edb74facc20f08f69a478c25e6fc2d207d9846/.github/workflows/cpp_lint.yml#L4).
> 3. Regarding the CLA issue, I believe someone from MLCommons needs to add you. @tushar-krishna or @srinivas212 should be able to help with this.
> 
> I'll run some tests and share the results.

Hi @JoongunPark ,
Thanks for the feedback.
1. Merge conflicts have been fixed.
2. I checked cpp_lint.yml in .github/workflows and it is exactly the same as in the main branch.

**Comment by JoongunPark** [comment]
Hi @rvinaybharadwaj,

Thank you for updating your PR!
I have two last requests:

1. Could you please ensure that the "CLA" test passes? It's mandatory, and unfortunately, I can't assist with this part.
2. It seems you've successfully addressed the original issue with the C++ Link (the module installation problem). However, there are still a few minor errors. You can check them here: [link](https://github.com/mlcommons/chakra/actions/runs/10746747873/job/29808124017?pr=145). Could you please take a look and fix those as well?

Thank you for your efforts!

**Comment by rvinaybharadwaj** [comment]
> Hi @rvinaybharadwaj,
> 
> Thank you for updating your PR! I have two last requests:
> 
> 1. Could you please ensure that the "CLA" test passes? It's mandatory, and unfortunately, I can't assist with this part.
> 2. It seems you've successfully addressed the original issue with the C++ Link (the module installation problem). However, there are still a few minor errors. You can check them here: [link](https://github.com/mlcommons/chakra/actions/runs/10746747873/job/29808124017?pr=145). Could you please take a look and fix those as well?
> 
> Thank you for your efforts!

Hi @JoongunPark,
All errors have been fixed.

**Comment by JoongunPark** [comment]
@rvinaybharadwaj 
Thank you for resolving all the issues! I've started reviewing your code in detail, and here’s my understanding so far:

### 1. **New Classes**
   - **`JSONNode`:** Handles execution trace (ET) parsing from JSON files.
   - **`WrapperNode`:** Abstracts ET handling for multiple formats (e.g., `.et` and `.json`), working within the ETFeeder component.

### 2. **Dependency Changes**
   - **`pyproject.toml`:** Updated to support `protobuf==5.*`. This resolves some installation conflicts with the original version, ensuring compatibility.

### 3. **Test Coverage**
   - **Unit Tests (tests/feeder/wrapperTests.cpp):** Verifies the JSON parsing functionality.
   - **Test Data (tests/data):** Includes new JSON-format test files.

### 4. **Linting Updates**
   - **clang-format-lint:** Clang format (11->16). Linter version (v0.11 -> v0.18.1)

### 5. **New Third-Party Library**
   - **`json.hpp` (src/third_party/utils/json.hpp):** Utilized by `JSONNode` for JSON parsing.
   
Also, I have a question regarding JSON format trace. This PR does not include trace conversion from ET to JSON. 
Is this PR assuming to get JSON trace with exsting JSONizer? 

Thank you again!

**Comment by rvinaybharadwaj** [comment]
> @rvinaybharadwaj Thank you for resolving all the issues! I've started reviewing your code in detail, and here’s my understanding so far:
> 
> ### 1. **New Classes**
> * **`JSONNode`:** Handles execution trace (ET) parsing from JSON files.
> * **`WrapperNode`:** Abstracts ET handling for multiple formats (e.g., `.et` and `.json`), working within the ETFeeder component.
> 
> ### 2. **Dependency Changes**
> * **`pyproject.toml`:** Updated to support `protobuf==5.*`. This might resolve some installation conflicts with the original version.
> 
> ### 3. **Test Coverage**
> * **Unit Tests (tests/feeder/wrapperTests.cpp):** Verifies the JSON parsing functionality.
> * **Test Data (tests/data):** Includes new JSON-format test files.
> 
> ### 4. **Linting Updates**
> * **clang-format-lint:** Clang format (11->16). Linter version (v0.11 -> v0.18.1)
> 
> ### 5. **New Third-Party Library**
> * **`json.hpp` (src/third_party/utils/json.hpp):** Utilized by `JSONNode` for JSON parsing.
> 
> Also, I have a question regarding JSON format trace. This PR does not include trace conversion from ET to JSON. Is this PR assuming to get JSON trace with exsting Jsonizer?
> 
> Thank you again!

@JoongunPark
Thank you for reviewing the code and summarizing. Your understanding of different parts of the code is correct. The existing JSONizer outputs a slightly different format, but I was assuming it easy to do a JSON dump directly from Pytorch or any other tool.

**Comment by JoongunPark** [comment]
> > @rvinaybharadwaj Thank you for resolving all the issues! I've started reviewing your code in detail, and here’s my understanding so far:
> > ### 1. **New Classes**
> > 
> > * **`JSONNode`:** Handles execution trace (ET) parsing from JSON files.
> > * **`WrapperNode`:** Abstracts ET handling for multiple formats (e.g., `.et` and `.json`), working within the ETFeeder component.
> > 
> > ### 2. **Dependency Changes**
> > 
> > * **`pyproject.toml`:** Updated to support `protobuf==5.*`. This might resolve some installation conflicts with the original version.
> > 
> > ### 3. **Test Coverage**
> > 
> > * **Unit Tests (tests/feeder/wrapperTests.cpp):** Verifies the JSON parsing functionality.
> > * **Test Data (tests/data):** Includes new JSON-format test files.
> > 
> > ### 4. **Linting Updates**
> > 
> > * **clang-format-lint:** Clang format (11->16). Linter version (v0.11 -> v0.18.1)
> > 
> > ### 5. **New Third-Party Library**
> > 
> > * **`json.hpp` (src/third_party/utils/json.hpp):** Utilized by `JSONNode` for JSON parsing.
> > 
> > Also, I have a question regarding JSON format trace. This PR does not include trace conversion from ET to JSON. Is this PR assuming to get JSON trace with exsting Jsonizer?
> > Thank you again!
> 
> @JoongunPark Thank you for reviewing the code and summarizing. Your understanding of different parts of the code is correct. The existing JSONizer outputs a slightly different format, but I was assuming it is easy to do a JSON dump directly from Pytorch or any other tool.

Regarding the JSON format trace, I believe we should discuss whether to align it with the existing JSONizer or consider using a different format. This decision might be crucial, as I recall your initial plan was to create a merged, unified JSON trace. We can also involve @TaekyungHeo for further discussion on this matter.

File: .github/workflows/cpp_lint.yml [hunk 1]
```diff
@@ -11,9 +11,9 @@ jobs:
       uses: actions/checkout@v2
 
     - name: Format and Lint C++ Code
-      uses: DoozyX/clang-format-lint-action@v0.11
+      uses: DoozyX/clang-format-lint-action@v0.18.1
```
Comment (TaekyungHeo) [inline .github/workflows/cpp_lint.yml:14]: Could you please create a separate PR for updating `.github/workflows/cpp_lint.yml`? Why is updating the versions necessary for this PR?

File: pyproject.toml [hunk 2]
```diff
@@ -11,7 +11,7 @@ license = {file = "LICENSE.md"}
 authors = [
     {name = "MLCommons", email = "chakra@mlcommons.org"},
 ]
-dependencies = ["protobuf==4.*", "graphviz", "networkx", "pydot"]
+dependencies = ["protobuf==5.*", "graphviz", "networkx", "pydot"]
```
Comment (TaekyungHeo) [inline pyproject.toml:14]: Could you please explain why updating the protobuf version was necessary?

File: src/feeder/JSONNode.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
+#include "JSONNode.h"
```
Comment (TaekyungHeo) [inline src/feeder/JSONNode.cpp:1]: Let's be consistent with naming. The existing file names follow snake case.

File: src/feeder/JSONNode.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
+#pragma once
```
Comment (TaekyungHeo) [inline src/feeder/JSONNode.h:1]: Let's be consistent with naming. The existing file names follow snake case.

File: src/feeder/WrapperNode.cpp [hunk 5]
```diff
@@ -0,0 +1,645 @@
+#include "WrapperNode.h"
```
Comment (TaekyungHeo) [inline src/feeder/WrapperNode.cpp:1]: Let's be consistent with naming. The existing file names follow snake case.

File: src/feeder/et_feeder.cpp [hunk 6]
```diff
@@ -173,4 +173,4 @@ void ETFeeder::readNextWindow() {
       dep_free_node_queue_.emplace(node);
     }
   }
-}
+}
```
Comment (TaekyungHeo) [inline src/feeder/et_feeder.cpp:176]: This change is minor, but it's unnecessary. Please avoid introducing unnecessary changes that are unrelated to this PR.

File: tests/feeder/wrapperTests.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    node.addNode(jnode1);
+    JSONNode jnode2;
+    node.lookupNode(216);
+    jnode2 = node.getJSONNode();
+    ASSERT_EQ(jnode2.id(), 216);
+  }
+}
+
+TEST_F(WrapperNodeTest, NodeGetChildrenTest) {
+  // std::string filename = "tests/data/chakra.0.et";
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:154]: Please don't comment out codes.

**Review by TaekyungHeo — COMMENTED** [review]
I left several comments and will leave more when I have more time.

File: src/feeder/JSONNode.cpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  node_id = t.node_id;
+  node_name = t.node_name;
+  node_type = t.node_type;
+  is_cpu_op = t.is_cpu_op;
+  runtime = t.runtime;
+  data_deps = t.data_deps;
+  dep_unresolved_parent_ids_json = t.dep_unresolved_parent_ids_json;
+  children_vec_json = t.children_vec_json;
+  children_set_json = t.children_set_json;
+  /* 	COMM_SEND_NODE = 5,
```
Comment (TaekyungHeo) [inline src/feeder/JSONNode.cpp:17]: Let's use the autogenerated C++ files and variables in the code, rather than hardcoding values directly.

File: src/feeder/JSONNode.cpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    runtime = data["workload_graph"][id]["runtime"];
+  } catch (...) {
+  }
+  try {
+    data_deps =
+        data["workload_graph"][id]["data_deps"].get<std::vector<uint64_t>>();
+  } catch (...) {
+    std::cerr << "data deps not specified in ET" << std::endl;
+  }
+  /* 	COMM_SEND_NODE = 5,
```
Comment (TaekyungHeo) [inline src/feeder/JSONNode.cpp:63]: Let's use the autogenerated C++ files and variables in the code, rather than hardcoding values directly.

File: tests/data/jsonData.tar.gz
Comment (TaekyungHeo) [inline tests/data/jsonData.tar.gz:1]: Let's be consistent with naming. The existing file names follow snake case.

File: tests/feeder/tests.cpp [hunk 10]
```diff
@@ -136,4 +136,4 @@ TEST_F(ETFeederTest, NodeGetChildrenTest) {
 int main(int argc, char** argv) {
   ::testing::InitGoogleTest(&argc, argv);
   return RUN_ALL_TESTS();
-}
+}
```
Comment (TaekyungHeo) [inline tests/feeder/tests.cpp:139]: This change is minor, but it's unnecessary. Please avoid introducing unnecessary changes that are unrelated to this PR.

File: tests/feeder/wrapperTests.cpp [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+  virtual void TearDown() {
+    node.releaseMemory();
+  }
+
+  WrapperNode node;
+};
+
+TEST_F(WrapperNodeTest, ConstructorNodeIDTest) {
+  // SetUp("tests/data/chakra.0.et");
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:21]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+  uint64_t firstNodeID = node.getNodeID();
+  ASSERT_EQ(firstNodeID, 216);
+
+  node.getNextIssuableNode();
+  uint64_t secondNodeID = node.getNodeID();
+  ASSERT_EQ(secondNodeID, 432);
+}
+
+TEST_F(WrapperNodeTest, ConstructorNodeValuesTest) {
+  // SetUp("tests/data/chakra.0.et");
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:33]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+  ASSERT_TRUE(node.isCPUOp());
+
+  node.getNextIssuableNode();
+  uint64_t secondNodeType = node.getNodeType();
+  ASSERT_EQ(secondNodeType, ChakraProtoMsg::COMM_COLL_NODE);
+  ASSERT_TRUE(node.isCPUOp());
+}
+
+TEST_F(WrapperNodeTest, ConstructorWrapperNodeTest) {
+  // std::string filename = "tests/data/chakra.0.et";
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:47]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    std::vector<JSONNode> children;
+    node.getChildren(children);
+    ASSERT_EQ(children[0].id(), 217);
+    ASSERT_EQ(children[1].id(), 430);
+    ASSERT_EQ(children[2].id(), 435);
+  }
+}
+
+TEST_F(WrapperNodeTest, RemoveTest) {
+  // SetUp("tests/data/chakra.0.et");
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:68]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    node.lookupNode(216);
+    ASSERT_TRUE(false) << "node should be removed \n";
+  } catch (const std::exception& e) {
+    // this is the desired behaviour
+  }
+  freopen("/dev/tty", "w", stderr);
+}
+
+TEST_F(WrapperNodeTest, RemoveAndGetNextTest) {
+  // SetUp("tests/data/chakra.0.et");
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:84]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+  SetUp("tests/data/small_chakra.0.json");
+  node.lookupNode(216);
+  ASSERT_EQ(node.getNodeID(), 216);
+  node.removeNode(216);
+  node.getNextIssuableNode();
+  ASSERT_EQ(node.getNodeID(), 216);
+}
+
+TEST_F(WrapperNodeTest, FreeChildrenTest) {
+  // SetUp("tests/data/chakra.0.et");
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:94]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+  ASSERT_EQ(node.getNodeID(), 216);
+  node.freeChildrenNodes(216);
+  node.getNextIssuableNode();
+  ASSERT_EQ(node.getNodeID(), 216);
+  node.getNextIssuableNode();
+  ASSERT_EQ(node.getNodeID(), 217);
+}
+
+TEST_F(WrapperNodeTest, HasNodesToIssueTest) {
+  // SetUp("tests/data/chakra.0.et");
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:106]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+  SetUp("tests/data/small_chakra.0.json");
+  node.getNextIssuableNode();
+  ASSERT_EQ(node.getNodeID(), 216);
+  ASSERT_TRUE(node.hasNodesToIssue());
+  node.removeNode(5);
+  ASSERT_TRUE(node.hasNodesToIssue());
+}
+
+TEST_F(WrapperNodeTest, PushBackIssuableNodeTest) {
+  // SetUp("tests/data/chakra.0.et");
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:116]: Please don't comment out codes.

File: tests/feeder/wrapperTests.cpp [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+  SetUp("tests/data/small_chakra.0.json");
+  node.pushBackIssuableNode(217);
+  node.getNextIssuableNode();
+  ASSERT_EQ(node.getNodeID(), 216);
+  node.getNextIssuableNode();
+  ASSERT_EQ(node.getNodeID(), 217);
+}
+
+TEST_F(WrapperNodeTest, AddNodeTest) {
+  // std::string filename = "tests/data/chakra.0.et";
```
Comment (TaekyungHeo) [inline tests/feeder/wrapperTests.cpp:126]: Please don't comment out codes.

File: src/feeder/et_feeder.h [hunk 20]
```diff
@@ -33,13 +33,12 @@ class ETFeeder {
   void pushBackIssuableNode(uint64_t node_id);
   std::shared_ptr<ETFeederNode> lookupNode(uint64_t node_id);
   void freeChildrenNodes(uint64_t node_id);
-
- private:
   void readGlobalMetadata();
   std::shared_ptr<ETFeederNode> readNode();
   void readNextWindow();
   void resolveDep();
```
Comment (TaekyungHeo) [inline src/feeder/et_feeder.h:39]: What was the reason for changing them from private to public?

File: .github/workflows/cpp_lint.yml (hunk 1 above)
Comment (rvinaybharadwaj) [reply .github/workflows/cpp_lint.yml:14]: The CPP linter was erroring out with
``` 
Traceback (most recent call last):
  File "/run-clang-format.py", line 25, in <module>
    from distutils.util import strtobool
ModuleNotFoundError: No module named 'distutils'
```
You can see that other outstanding PRs are also failing with the same error. I changed the version in order to fix the error.

File: pyproject.toml (hunk 2 above)
Comment (rvinaybharadwaj) [reply pyproject.toml:14]: There were some installation conflicts with protobuf version 4. 
```
/opt/hostedtoolcache/Python/3.10.14/x64/lib/python3.10/site-packages/_pytest/assertion/rewrite.py:178: in exec_module
    exec(co, module.__dict__)
tests/jsonizer/test_jsonizer.py:5: in <module>
    from chakra.schema.protobuf.et_def_pb2 import GlobalMetadata
/opt/hostedtoolcache/Python/3.10.14/x64/lib/python3.10/site-packages/chakra/schema/protobuf/et_def_pb2.py:9: in <module>
    from google.protobuf import runtime_version as _runtime_version
E   ImportError: cannot import name 'runtime_version' from 'google.protobuf' (/opt/hostedtoolcache/Python/3.10.14/x[64](https://github.com/mlcommons/chakra/actions/runs/10691857371/job/29680367619#step:7:65)/lib/python3.10/site-packages/google/protobuf/__init__.py)
```

This can also be seen in other outstanding PRs in the repo. Changing to protobuf 5 fixed the issue.

File: tests/feeder/wrapperTests.cpp (hunk 7 above)
Comment (rvinaybharadwaj) [reply tests/feeder/wrapperTests.cpp:154]: I have two input files for testing: `tests/data/chakra.0.json` and `tests/data/small_chakra.0.json`. small_chakra.0.json is a pruned version of the former for quick testing. I left a comment there to indicate the availability of the full dataset.

File: tests/data/jsonData.tar.gz
[the remark this answers is not in the document]
Comment (rvinaybharadwaj) [reply tests/data/jsonData.tar.gz:1]: I used a python script to convert the protobuf ET in tests/data to JSON.

File: src/feeder/et_feeder.h (hunk 20 above)
Comment (rvinaybharadwaj) [reply src/feeder/et_feeder.h:39]: I had to make them public to use them in the wrapper.

File: src/feeder/JSONNode.cpp (hunk 8 above)
Comment (rvinaybharadwaj) [reply src/feeder/JSONNode.cpp:17]: The enum if defined in the autogenerated protobuf header and I did not want to add protobuf dependency to JSON node. I can replicate the enumerated data structure in the JSON node header and use those variables for readability.

File: src/feeder/et_feeder.cpp (hunk 6 above)
Comment (rvinaybharadwaj) [reply src/feeder/et_feeder.cpp:176]: It was necessary to pass linter tests with the new linter version.

**Comment by rvinaybharadwaj** [comment]
Based on the discussion in the Chakra meeting on Dec 12, 2024 -

1. I check for the file format in [wrapper_node.cpp #L26](https://github.com/rvinaybharadwaj/chakra/blob/f803f335d14a40b517b6abba41b773fbba3bfa30/src/feeder/wrapper_node.cpp#L26) . Since, I use `filename.find_last_of(".")`, renaming the JSON file to _filename_.et.json should work without any changes.
2. In the future, if the protobuf ET files are renamed to _filename_.et.pb, we need to make the corresponding change in [wrapper_node.cpp #L27](https://github.com/rvinaybharadwaj/chakra/blob/f803f335d14a40b517b6abba41b773fbba3bfa30/src/feeder/wrapper_node.cpp#L27)

**Review by tushar-krishna — APPROVED** [review]
Thank you Vinay for the PR, and to Taekyung, Joongun and William for the review.
This looks good to merge.
